posts/2022-09-22-Rust-1.64.0.md

```diff
@@ -0,0 +1,256 @@
+over the network:
+
+```rust
+pub struct Error { ... }
+pub struct StorageResponse { ... }:
+pub struct StorageRequest(bool);
+
+impl StorageRequest {
+    /// Create a new instance of `StorageRequest`.
+    pub fn new() -> Self { ... }
+    /// Decide whether debug mode should be enabled.
+    pub fn set_debug(self, b: bool) -> Self { ... }
+    /// Send the request and receive a response.
+    pub async fn send(self) -> Result<StorageResponse, Error> { ... }
+}
+```
+
+Typical usage would likely look something like this:
+
+```rust
+let response = StorageRequest::new()  // 1. create a new instance
+    .set_debug(true)                  // 2. set some option
+    .send()                           // 3. construct the future
+    .await?;                          // 4. run the future + propagate errors
+```
+
+This is not bad, but we can do better here. Using `IntoFuture` we can combine
+_"construct the future"_ (line 3) and _"run the future"_ (line 4) into a single
+step:
+```rust
+let response = StorageRequest::new()  // 1. create a new instance
+    .set_debug(true)                  // 2. set some option
+    .await?;                          // 3. construct + run the future + propagate errors
+```
+
+We can do this by implementing `IntoFuture` for `StorageRequest`. `IntoFuture`
+requires us to have a named future we can return, which we can do by creating a
+"boxed future" and defining a type alias for it:
+
+```rust
+// First we must import some new types into the scope.
+use std::pin::Pin;
+use std::future::{Future, IntoFuture};
+
+pub struct Error { ... }
+pub struct StorageResponse { ... }
+pub struct StorageRequest(bool);
+
+impl StorageRequest {
+    /// Create a new instance of `StorageRequest`.
+    pub fn new() -> Self { ... }
+    /// Decide whether debug mode should be enabled.
+    pub fn set_debug(self, b: bool) -> Self { ... }
+    /// Send the request and receive a response.
+    pub async fn send(self) -> Result<StorageResponse, Error> { ... }
+}
+
+// The new implementations:
+// 1. create a new named future type
+// 2. implement `IntoFuture` for `StorageRequest`
+pub type StorageRequestFuture = Pin<Box<dyn Future<Output = Result<StorageResponse, Error> + Send + 'static>>
+impl IntoFuture for StorageRequest {
+    type IntoFuture = StorageRequestFuture;
+    type Output = <StorageRequestFuture as Future>::Output;
+    fn into_future(self) -> Self::IntoFuture {
+        Box::pin(self.send())
+    }
+}
+```
+This takes a bit more code to implement, but provides a simpler API for users.
+
+In the future, the Rust Async WG hopes to simplify the creating new named
+futures by supporting [`impl Trait` in `type` aliases (Type Alias Impl Trait or
+TAIT)](https://rust-lang.github.io/impl-trait-initiative/explainer/tait.html).
+This should make implementing `IntoFuture` easier by simplifying the type
+alias' signature, and make it more performant by removing the `Box` from the
+type alias.
+
+### Stabilizing `ready!` and `poll_fn`
```

Comment:
Something about formatting went wrong -- not quite sure where -- but probably worth checking the generated content even if github's editor highlighting isn't fixed.

Reply:
The generated content looks correct, and vim's highlighting seems fine with it too. This seems like a bug in github's highlighting.